### ISSUE TITLE
Separate DB inputs namespace in ProcessBuilder

### ISSUE THREAD

**Issue by giovannipizzi** [aiidateam/aiida_core#1532]
How much work do you think it would be to move all actual inputs (DB links) into a sub-namespace of the process builder like the one below?
The implementation below already works for the Builder, but I’m not sure how complicated this would be to actually make it work when we pass it down to the process.

The reason would be to avoid to cover a few keywords in the input namespace (like options, label, description, …) that a plugin developer might want to use. Of course, if needed, the changes should be done also at the lower lower levels (e.g. Process) so that one can actually e.g. define an input called `options`.

Mentioning also @greschd for feedback

```diff
diff --git a/aiida/work/job_processes.py b/aiida/work/job_processes.py
index c37b8861d..d88a84dcc 100644
--- a/aiida/work/job_processes.py
+++ b/aiida/work/job_processes.py
@@ -344,6 +344,7 @@ class JobProcess(processes.Process):
     TRANSPORT_OPERATION = 'TRANSPORT_OPERATION'
     CALC_NODE_LABEL = 'calc_node'
     OPTIONS_INPUT_LABEL = 'options'
+    INPUTS_LABEL = 'inputs'
 
     _calc_class = None
 
@@ -360,7 +361,8 @@ class JobProcess(processes.Process):
             super(JobProcess, cls_).define(spec)
 
             # Define the 'options' inputs namespace and its input ports
-            spec.input_namespace(cls.OPTIONS_INPUT_LABEL, help='various options')
+            spec.input_namespace(cls.OPTIONS_INPUT_LABEL, help='Various options')
+            spec.input_namespace(cls.INPUTS_LABEL, help='Calculation input links')
             spec.input('{}.resources'.format(cls.OPTIONS_INPUT_LABEL), valid_type=dict,
                 help='Set the dictionary of resources to be used by the scheduler plugin, like the number of nodes, '\
                      'cpus etc. This dictionary is scheduler-plugin dependent. Look at the documentation of the scheduler.')
@@ -403,9 +405,9 @@ class JobProcess(processes.Process):
                 additional_parameter = use_method.get('additional_parameter')
 
                 if additional_parameter:
-                    spec.input_namespace(key, help=docstring, valid_type=valid_type, required=False, dynamic=True)
+                    spec.input_namespace("{}.{}".format(cls.INPUTS_LABEL, key), help=docstring, valid_type=valid_type, required=False, dynamic=True)
                 else:
-                    spec.input(key, help=docstring, valid_type=valid_type, required=False)
+                    spec.input("{}.{}".format(cls.INPUTS_LABEL,key), help=docstring, valid_type=valid_type, required=False)
 
             # Outputs
             spec.outputs.valid_type = Data
```

**Comment by greschd** [comment]
I don't think you can cleanly separate inputs which are DB links from those that are not. Specifically, if you have a namespace that contains the inputs for a sub-process, it will typically contain both kinds of input.

Of course you could still create a separate namespace for the "user-created" inputs vs. inputs defined by AiiDA. However, I'm not sure there's a nice way to then hide the additional nesting level from the end user. Right now, this would just mean that the plugin developer has to create all inputs with 

```
spec.input('input.input_name')
```
instead of 
```
spec.input('input_name')
```

While I see the appeal of not having to reserve some names, I think on balance it wouldn't make things easier to use.

**Comment by sphuber** [comment]
I have to agree with @greschd here. I think the trade-off decision here, between having no "reserved" slots in the inputs dictionary, vs having a single way off defining both the process inputs as well as the more attribute like inputs such as the description and label, it favors the latter. If you cordon off those few attributes (`description`, `label, `options` and `store_provenance`) into their own space relative to the actual inputs, you incur a whole lot of extra complexity, both in using the code as well as developing and maintaining it. All at the gain of users having access to those keywords. Where I really think that only `options` stands a chance for users having a valid argument of wanting it to use for something else and I think they can overcome that limitation

**Comment by giovannipizzi** [comment]
Ok. But then e.g. there is also `submit_test` for the builder... Anyway, can we then at least raise an exception if the user tries to use, in the `specs`, one of these reserved keywords? If this list can be obtained programmatically and not as a hardcoded string, it would be ideal. Good if this is also mentioned in the docs.

**Comment by muhrin** [comment]
I'm also of the opinion that it's not necessarily an improvement in terms of the user interface to seperate these.

One thing I propose (which still has the 'shadowing' problems) is to add explicit 'metadata' i.e. things related to how the calculation was ran, etc not not an input.  The syntax could be
`builder.with(label=..., description=...)`
@giovannipizzi We discussed this yesterday and you objected (very reasonably) about the lack of tab completion, but I think we can implement this (functions have tab completion for arguments so if there is a runtime way to introduce function arguments we could do it - maybe dynamic function creation).

TL; DR:
Seperate out the 'metadata' and leave the inputs as they are.

**Comment by sphuber** [comment]
Closing in favor of duplicate #2180
